server/src/main/java/org/apache/druid/server/QueryScheduler.java

```diff
@@ -0,0 +1,226 @@
+   * anything acquired is also released.
+   *
+   * In the meantime, if a {@link ListenableFuture} is registered for the query that calls this method, it MUST handle
+   * this synchronization itself to ensure that no {@link Bulkhead} is acquired without releasing it.
+   */
+  public <T> Sequence<T> run(Query<?> query, Sequence<T> resultSequence)
+  {
+    List<Bulkhead> bulkheads = acquireLanes(query);
+    return resultSequence.withBaggage(() -> releaseLanes(bulkheads));
+  }
+
+  /**
+   * Forcibly cancel all futures that have been registered to a specific query id
+   */
+  public boolean cancelQuery(String id)
+  {
+    // if you re-use queryId and cancel queries... you are going to have a bad time
+    queryDatasources.removeAll(id);
+    Set<ListenableFuture<?>> futures = queryFutures.removeAll(id);
+    boolean success = true;
+    for (ListenableFuture<?> future : futures) {
+      success = success && future.cancel(true);
+    }
+    return success;
+  }
+
+  /**
+   * Get a {@link Set} of datasource names for a {@link Query} id, used by {@link QueryResource#cancelQuery} to
+   * authorize that a user may call {@link #cancelQuery} for the given id and datasources
+   */
+  public Set<String> getQueryDatasources(final String queryId)
+  {
+    return queryDatasources.get(queryId);
+  }
+
+  /**
+   * Get the maximum number of concurrent queries that {@link #run} can support
+   */
+  public int getTotalAvailableCapacity()
+  {
+    return laneRegistry.getConfiguration(TOTAL)
+                       .map(config -> laneRegistry.bulkhead(TOTAL, config).getMetrics().getAvailableConcurrentCalls())
+                       .orElse(-1);
+  }
+
+  /**
+   * Get the maximum number of concurrent queries that {@link #run} can support for a given lane
+   */
+  public int getLaneAvailableCapacity(String lane)
+  {
+    return laneRegistry.getConfiguration(lane)
+                       .map(config -> laneRegistry.bulkhead(lane, config).getMetrics().getAvailableConcurrentCalls())
+                       .orElse(-1);
+  }
+
+  /**
+   * Acquire a semaphore for both the 'total' and a lane, if any is associated with a query
+   */
+  @VisibleForTesting
+  List<Bulkhead> acquireLanes(Query<?> query)
+  {
+    final String lane = QueryContexts.getLane(query);
+    final Optional<BulkheadConfig> laneConfig = lane == null ? Optional.empty() : laneRegistry.getConfiguration(lane);
+    List<Bulkhead> hallPasses = new ArrayList<>(2);
+    final Optional<BulkheadConfig> totalConfig = laneRegistry.getConfiguration(TOTAL);
+    // if we have a lane, get it first
+    laneConfig.ifPresent(config -> {
+      Bulkhead laneLimiter = laneRegistry.bulkhead(lane, config);
+      if (!laneLimiter.tryAcquirePermission()) {
+        throw new QueryCapacityExceededException(lane);
+      }
+      hallPasses.add(laneLimiter);
+    });
+
+    // everyone needs to take one from the total lane; to ensure we don't acquire a lane and never release it, we want
```

Comment:
What's the purpose of the TOTAL bulkhead if each lane has a bulkhead?

Reply:
the total is to control overall query throughput, and is tied to `druid.query.scheduler.numThreads`. Even queries without a lane are subject to this limit, if configured. I'll try to make this more clear in the javadocs or comments.

Reply:
I was thinking that if there are only counters per lane, then the code would be simpler as the lane and total counter would not have to be kept consistent.

Reply:
that is true, but I wanted to dual purpose this to also be able to set total limits to protect the rest of the service from the query processing system, and save room for health checks and the like. 

As I discussed in the description, I considered that the total counter specifically might be suitable as a jetty `QoSFilter`, which also is using a semaphore, however we would need to ensure that we set the request timeout high enough to give the laned queries time to fail and release the semaphore since the total would now be getting acquired before the lane. It seemed easier to me to track here than to precisely determine that interval, however the discussion is worth having I think.